backend/src/main/java/com/google/sps/FindMatchQuery.java

```diff
@@ -58,23 +51,39 @@ public Match findMatch(List<Participant> participants, Participant newParticipan
       int currDuration = currParticipant.getDuration();
       boolean compatibleDuration = Math.abs(newDuration - currDuration) <= MAX_DURATION_DIFF;
       int duration = Math.min(newDuration, currDuration);
-      
+
       if (!compatibleDuration) {
         continue;
       }
-      
+
       // Check if participants are both free for that duration + extra
-      long newTimeAvailableUntil = newParticipant.getTimeAvailableUntil();
-      long currTimeAvailableUntil = currParticipant.getTimeAvailableUntil();
-      boolean compatibleTime = Instant.now(clock).plusMillis(Duration.ofMinutes(duration + PADDING_TIME).toMillis())
-          .isBefore(Instant.ofEpochMillis(Math.min(newTimeAvailableUntil, currTimeAvailableUntil)));
+      ZonedDateTime newEndTimeAvailable = newParticipant.getEndTimeAvailable();
+      ZonedDateTime currEndTimeAvailable = currParticipant.getEndTimeAvailable();
+      ZonedDateTime earliestEndTimeAvailable =
+          getEarlier(newEndTimeAvailable, currEndTimeAvailable);
+      boolean compatibleTime =
+          dateTime.plusMinutes(duration + PADDING_TIME).isBefore(earliestEndTimeAvailable);
 
       if (compatibleTime) {
         // TODO: change match ID (currently -1 for easy error checking)
-        return new Match(/* id= */ -1L, /* firstParticipant= */ newParticipant, /* secondParticipant= */ currParticipant, /* duration= */ duration, /* timestamp= */ date.getTime());
+        return new Match(
+            /* id= */ -1L,
+            /* firstParticipant= */ newParticipant,
+            /* secondParticipant= */ currParticipant,
+            /* duration= */ duration,
+            /* timestamp= */ dateTime.toInstant().toEpochMilli());
```

Comment:
non-blocking: Adding these to all arguments is entirely ok. Internally we only require them for literals like `-1L` since named variables often include details about the corresponding parameter e.g. `duration` above is passed as the argument for the `duration` parameter which is already unambiguous.